bundles/org.openhab.binding.modbus.sunspec/src/main/java/org/openhab/binding/modbus/sunspec/internal/discovery/SunspecDiscoveryProcess.java

```diff
@@ -0,0 +1,397 @@
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryResultBuilder;
+import org.eclipse.smarthome.core.library.types.DecimalType;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.modbus.discovery.ModbusDiscoveryListener;
+import org.openhab.binding.modbus.handler.EndpointNotInitializedException;
+import org.openhab.binding.modbus.handler.ModbusEndpointThingHandler;
+import org.openhab.binding.modbus.sunspec.internal.dto.CommonModelBlock;
+import org.openhab.binding.modbus.sunspec.internal.dto.ModelBlock;
+import org.openhab.binding.modbus.sunspec.internal.parser.CommonModelParser;
+import org.openhab.io.transport.modbus.BasicModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.BasicPollTaskImpl;
+import org.openhab.io.transport.modbus.BitArray;
+import org.openhab.io.transport.modbus.ModbusBitUtilities;
+import org.openhab.io.transport.modbus.ModbusConstants.ValueType;
+import org.openhab.io.transport.modbus.ModbusReadCallback;
+import org.openhab.io.transport.modbus.ModbusReadFunctionCode;
+import org.openhab.io.transport.modbus.ModbusReadRequestBlueprint;
+import org.openhab.io.transport.modbus.ModbusRegisterArray;
+import org.openhab.io.transport.modbus.ModbusSlaveErrorResponseException;
+import org.openhab.io.transport.modbus.PollTask;
+import org.openhab.io.transport.modbus.endpoint.ModbusSlaveEndpoint;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class is used by the SunspecDiscoveryParticipant to detect
+ * the model blocks defined by the given device.
+ * It scans trough the defined model items and notifies the
+ * discovery service about the discovered devices
+ *
+ * @author Nagy Attila Gabor - Initial contribution
+ */
+@NonNullByDefault
+public class SunspecDiscoveryProcess {
+
+    /**
+     * Logger instance
+     */
+    private final Logger logger = LoggerFactory.getLogger(SunspecDiscoveryProcess.class);
+
+    /**
+     * The handler instance for this device
+     */
+    private final ModbusEndpointThingHandler handler;
+
+    /**
+     * The endpoint where we can reach the device
+     */
+    private final ModbusSlaveEndpoint endpoint;
+
+    /**
+     * Listener for the discovered devices. We get this
+     * from the main discovery service, and it is used to
+     * submit any discovered Sunspec devices
+     */
+    private final ModbusDiscoveryListener listener;
+
+    /**
+     * The endpoint's slave id
+     */
+    private int slaveId;
+
+    /**
+     * Number of maximum retries
+     */
+    private static final int maxTries = 3;
+
+    /**
+     * List of start addresses to try
+     */
+    private List<Integer> possibleAddresses;
```

Comment:
From the way you are using this field, it should be a Queue instead.
```suggestion
    private Queue<Integer> possibleAddresses;
```

Reply:
You are right, this suits the scenario much better.